Fill full height of context section in TUI mode

peace-maker opened the pull request:
Add new `context-tui-adjust-height` config option to dynamically increase the number of lines printed in `regs`, `disasm`, `code`, and `threads` sections to fill the whole TUI window allocated for the section.

This ignores the other config values in place to limit the number of lines for each individual section and interacts weirdly with the context history, because the same entry could be added twice with longer section output when switching from CLI to TUI mode and the context was already output in CLI mode. That is a future TODO to replace the section data in the latest history entry when applicable.

<img width="2532" height="1353" alt="grafik" src="https://github.com/user-attachments/assets/6e06b686-050d-44ba-8b13-8c9813dc5e57" />

<img width="2530" height="1349" alt="grafik" src="https://github.com/user-attachments/assets/58fbbb7b-c36d-4c34-b2a4-3cfe6fff98e7" />

This is part of #2183 but streching the disasm window in the CLI context output to use the full height isn't implemented yet.

peace-maker committed:
Fill full height of context section in TUI mode

Add new `context-tui-adjust-height` config option to dynamically increase
the number of lines printed in `regs`, `disasm`, `code`, and `threads` sections
to fill the whole TUI window allocated for the section.

This ignores the other config values in place to limit the number of lines
for each individual section and interacts weirdly with the context history,
because the same entry could be added twice with longer section output
when switching from CLI to TUI mode and the context was already output
in CLI mode. That is a future TODO to replace the section data in the latest
history entry when applicable.

k4lizen commented on pwndbg/gdblib/tui/context.py:
```diff
@@ -15,6 +15,17 @@
+    True,
+    "adjust height of context sections to fit the TUI",
+    help_docstring="""
+    Adjust the height of context sections to fit the TUI window, ignoring section specific limits.
```
other help_docstring's in the code look like this
```
    help_docstring="""
The bins are traversed both forwards and backwards.
""",
```
is the indentation here fine?

peace-maker replied:
Down the line the `help_docstring` is stripped, so there shouldn't be a difference.

https://github.com/pwndbg/pwndbg/blob/dfdfbd4f7bd45869d7b4b7d3e4418aede29359d2/pwndbg/lib/config.py#L241

k4lizen replied:
strip only deletes whitespace from the end and start of the string, you have indentation effectively in the middle of the string

k4lizen replied:
<img width="943" height="163" alt="image" src="https://github.com/user-attachments/assets/5ed2bbe0-0f03-4e63-87b0-5db7c19b6f6a" />

peace-maker replied:
Ah, right!

k4lizen commented on pwndbg/commands/context.py:
```diff
@@ -319,10 +319,17 @@ def output(section: str):
 
 @pwndbg.commands.Command(parser, aliases=["ctx-out"], category=CommandCategory.CONTEXT)
-def contextoutput(section, path, clearing, banner="both", width: int = None):
+def contextoutput(section, path, clearing, banner="both", width: int = None, height: int = None):
```
`height: int | None = None`

k4lizen commented on pwndbg/commands/context.py:
```diff
@@ -319,10 +319,17 @@ def output(section: str):
+    type=int,
+    nargs="?",
+    default=None,
+    help="Sets a fixed height (used for number of lines to display before cutoff). Set to None for auto",
```
maybe  extend the comment a bit to describe that it ignores specific limits, extends some contexts while not others etc

k4lizen commented on pwndbg/commands/context.py:
```diff
@@ -889,7 +898,7 @@ def compact_regs(regs: List[str], width=None, target=sys.stdout) -> List[str]:
 
 @serve_context_history
-def context_regs(target=sys.stdout, with_banner=True, width=None):
+def context_regs(target=sys.stdout, with_banner=True, width=None, height=None):
```
would be nice to add a type hint to the other places too

k4lizen reviewed (changes requested):
plz check the `help_docstring` thing, otherwise lgtm

disconnect3d commented:
docs on CI:
```
VERIFICATION FAILED. The files differ from what would be auto-generated.
Error: File docs/commands/context/contextoutput.md is too short. Expected 30 lines, got 29.
Please run ./scripts/generate-docs.sh from project root and commit the changes.
Error: Process completed with exit code 2.
```

peace-maker committed:
Improve help strings and regenerate docs

peace-maker committed:
Add typehints to context commands

peace-maker committed:
Add type hints for context output redirection

The CallOutput redirection doesn't fully implement the TextIO protocol but only the necessary functions used in the code. Instead of implementing all of them keep a type union around.

peace-maker commented:
I believe the failing tests are unrelated to the changes in this branch.

k4lizen reviewed (approved):
they are.

thanks, and for the drive-by type hint fixes as well :p

peace-maker committed:
Merge branch 'dev' into tui_full_height
